docs/versioning.md

```diff
@@ -0,0 +1,488 @@
+
+## Implementation of Bucket Versioning in S3
+
+### Roles of Metadata Engine and API Component
+
+Each version of an object is stored as a separate key in metadata. The API
+component of S3 manages the put and delete of version data, sends instructions
+to create, retrieve, overwrite or delete a version's metadata to the metadata
+backend, and tracks extra information about a version, such as whether it is a
+delete marker or a null version.
+
+The metadata backend is responsible for handling the create, overwrite and
+delete operations of version metadata upon receiving instructions. We use
+version identifiers as the suffix for the keys of the object versions, and a
+special version to represent the latest version.
+
+This is an example of what the keys may look like in metadata, for an object
+`foo/bar` with three versions. The latest version is represented below in the
+key `foo/bar`. This special version is called the master version and is
+described further in one of the following sections, "Master Version".
+
+| key |
+|-----|
+| foo/bar |
+| foo/bar\098506163554375999999PARIS  0.a430a1f85c6ec |
+| foo/bar\098506163554373999999PARIS  0.41b510cd0fdf8 |
+| foo/bar\098506163554373999998PARIS  0.f9b82c166f695 |
+
+The following is a summary of the design and implementation details of bucket
+versioning, described first for the metadata engine and then for the API.
+
+## Implementation of Bucket Versioning in Metadata
+
+### Version Id Format
+
+The version id is generated by the metadata backend, and encoded in a
+hexadecimal string format by S3. The following information is encoded before
+sending a response to a request. S3 also decodes the hexadecimal string
+received from a request before sending to metadata to retrieve a particular
+version.
+
+`object_name separator version_id`
+where:
+
+- `object_name`: is the key of the object in metadata
+- `separator`: we use the `null` character (`0x00` or `\0`) as the separator
+  between the `object_name` and the `version_id` of a key
+- `version_id`: is the version identifier; this encodes the ordering
+  information in the texture format as metadata orders keys alphabetically
+
+The format of a `version_id` is:
+`ts` `rep_group_id` `seq_id`
+where:
+
+- `ts`: is the combination of epoch and an increasing number
+- `rep_group_id`: is the name of deployment(s) considered one unit used for
+   replication later
+- `seq_id`: is supposed to be a unique value based on metadata information.
+
+An example of a key: `foo\01234567890123456PARIS 1234.123456` indicating that
+this specific version of `foo` was the `123456`th entry created during the epoch
+`1234567890` in the replication group `PARIS` with `1234.1234566` as `seq_id`.
+
+### Master Version
+
+We store a copy of the latest version of an object's metadata using
+`object_name` as the key; this version is called the master version. The master
+versions of the objects facilitate the standard GET operation, which would
```

Comment:
The master version of each object facilitates the standard GET and LIST operations, which would..